### ISSUE TITLE
Provide Wasm (and Rust) function to get valid principal/action/resource for a policy

### ISSUE THREAD

**Issue by cdisselkoen** [cedar-policy/cedar#801]
### Category

User level API features/changes

### Describe the feature you'd like to request

[vscode-cedar](https://github.com/cedar-policy/vscode-cedar)'s autocomplete functionality needs a function in Wasm which takes a policy and schema, and returns all of the valid principal types for that policy, all of the valid actions for that policy, and all of the valid resource types for that policy. 

The Rust API doesn't currently have this functionality either, so while we're at it, we should add this to the public Rust API too.

### Describe alternatives you've considered

As a hack/workaround, the current code creates special intentionally-invalid policies and parses the errors produced by those.

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [ ] ⚠️ This feature might incur a breaking change

**Comment by hakanson** [comment]
An example based on the vscode-cedar [testdata](https://github.com/cedar-policy/vscode-cedar/blob/prerelease/testdata/narrow/policies.cedar):

```
@id("E1,E2 a,a2 R2")
permit (principal, action, resource is NS::R2);
```

the constraints (policy export to JSON) are `All` for `principal` and `action`

```json
{
  "effect": "permit",
  "principal": {
    "op": "All"
  },
  "action": {
    "op": "All"
  },
  "resource": {
    "op": "is",
    "entity_type": "NS::R2"
  },
  "conditions": [],
  "annotations": {
    "id": "E1,E2 a,a2 R2"
  }
}
```

but looking at the namespace, they are:
- principal: `NS::E1`, `NS::E2`
- action: `NS::Action::"a"`, `NS::Action::"a2"`
- resource: `NS::R2`

```
namespace NS {
  entity E;
  entity R1 in [R] = {"p1": String};
  entity R;
  entity R2 in [R] = {"p1": Long};
  entity E1 in [E] = {"p1": String};
  entity E2 in [E] = {"p1": Long};
  action "as";
  action "a" in [Action::"as"] appliesTo {
    principal: [E1, E2],
    resource: [R1, R2],
    context: {"c1": Long}
  };
  action "a1" in [Action::"as"] appliesTo {
    principal: [E1],
    resource: [R1],
    context: {"c1": Long}
  };
  action "a2" in [Action::"as"] appliesTo {
    principal: [E2],
    resource: [R2],
    context: {"c1": Long}
  };
}
```

**Comment by shaobo-he-aws** [comment]
Could you clarify what validity mean for this feature? I was wondering what's needed is the set of cross-products that type check.

**Comment by cdisselkoen** [comment]
Yes, I think that's what's needed.  More specifically I think what @hakanson wants is "all principal types for which the policy could ever fire" and likewise for actions/resources.

**Comment by shaobo-he-aws** [comment]
> Yes, I think that's what's needed. More specifically I think what @hakanson wants is "all principal types for which the policy could ever fire" and likewise for actions/resources.

Got it. Thanks.

**Comment by hakanson** [comment]
example of the code I'd like to replace is `determineEntityTypes` in https://github.com/cedar-policy/vscode-cedar/blob/prerelease/src/validate.ts#L341-L377 with tests at https://github.com/cedar-policy/vscode-cedar/blob/prerelease/src/test/suite/validation.test.ts#L427-L481

**Comment by shaobo-he-aws** [comment]
Implemented via #1074
